src/Analyser/MutatingScope.php

```diff
@@ -569,6 +569,18 @@ private function resolveType(Expr $node): Type
 			return new BooleanType();
 		}
 
+		if ($node instanceof \PhpParser\Node\Expr\BitwiseNot) {
+			if ($this->treatPhpDocTypesAsCertain) {
+				$exprType = $this->getType($node->expr);
+			} else {
+				$exprType = $this->getNativeType($node->expr);
+			}
+			if ($exprType instanceof StringType) {
+				return new StringType();
+			}
+			return $exprType->toInteger();
```

Comment:
Not covered cases:
* floats
* unions of constant strings

Best way to tackle this is to use TypeTraverse::map() which takes care of complex types easily...

Reply:
Floats get coerced to integer for bitwise not. I hadn't considered the union types problem.

Reply:
Oh right, floats work (but a unit test would be nice).